src/prefect/engine/flow_runner.py

```diff
@@ -469,7 +564,8 @@ def get_flow_run_state(
             all_final_states = final_states.copy()
             for t, s in list(final_states.items()):
                 if s.is_mapped():
-                    s.map_states = executor.wait(s.map_states)
+                    # ensure we wait for any mapped children to complete
+                    s.map_states = executor.wait(mapped_children.get(t) or s.map_states)
```

Comment:
Why the `or s.map_states`? We don't want to store pending futures on the state, since that may be passed around via the executor, and sending futures between clients is error prone and expensive. For clarity I'd suggest:
```python
if t in mapped_children:
    s.map_states = executor.wait(mapped_children[t])
```

Reply:
At this step there are no further submissions through the executor, and setting this attribute is for user-facing interactions only (e.g., so users can access the states of the spawned tasks).  We need to reference `s.map_states` because this could be a re-run of a partially completed pipeline where we didn't actually submit any new work, and consequently don't want to lose the already resolved state information that `map_states` contains.

Reply:
> and consequently don't want to lose the already resolved state information that map_states contains.

Isn't that information  already in `s.mapped_states` (the existing logic ends up being `s.mapped_states = s.mapped_states` if `t` isn't in `mapped_children`)?

I'm just trying to clarify in the code that no futures are ever in `s.mapped_states`, so there's no point in calling `executor.wait` on them.

Reply:
Oh I see I see - yea that makes sense; will update.